File: vue-frontend/src/stores/aiChat.ts
import { defineStore } from 'pinia';
import { ref, watch } from 'vue';
import axios from 'axios';
import { ElMessage } from 'element-plus';
import { useTranslatorStore } from './translator';

export interface ChatMessage {
  id: string;
  role: 'user' | 'assistant' | 'system';
  content: string;
  timestamp: number;
}

export interface ApplyFieldInfo {
  key: string;
  path: string;
  label: string;
  value: any;
  preview: string; // truncated preview
}

const CHAT_STORAGE_KEY = 'aiChatMessages';
const CHAT_SIDEBAR_VISIBLE_KEY = 'aiChatSidebarVisible';

export const useAIChatStore = defineStore('aiChat', () => {
  // --- State ---
  const messages = ref<ChatMessage[]>([]);
  const isLoading = ref(false);
  const sidebarVisible = ref(false);

  // --- Actions ---
  const generateId = () => `msg_${Date.now()}_${Math.random().toString(36).slice(2, 8)}`;

  const loadFromStorage = () => {
    const savedMessages = localStorage.getItem(CHAT_STORAGE_KEY);
    if (savedMessages) {
      try {
        messages.value = JSON.parse(savedMessages);
      } catch (e) {
        localStorage.removeItem(CHAT_STORAGE_KEY);
      }
    }
    const savedVisible = localStorage.getItem(CHAT_SIDEBAR_VISIBLE_KEY);
    if (savedVisible) {
      sidebarVisible.value = savedVisible === 'true';
    }
  };

  const toggleSidebar = () => {
    sidebarVisible.value = !sidebarVisible.value;
  };

  const sendMessage = async (content: string) => {
    if (!content.trim()) return;

    const translatorStore = useTranslatorStore();
    const settings = translatorStore.translationSettings;

    if (!settings.api_key) {
      ElMessage.warning('请先在设置中提供您的 API Key');
      return;
    }

    // Add user message
    const userMessage: ChatMessage = {
      id: generateId(),
      role: 'user',
      content: content.trim(),
      timestamp: Date.now(),
    };
    messages.value.push(userMessage);

    // Send to API
    isLoading.value = true;
    try {
      const response = await axios.post('/api/v1/character/ai-chat', {
        messages: messages.value
          .filter(m => m.role !== 'system')
          .map(m => ({ role: m.role, content: m.content })),
        settings: {
          api_key: settings.api_key,
          base_url: settings.base_url,
          model_name: settings.model_name,
        },
        character_card: translatorStore.characterCard,
      });

      const assistantMessage: ChatMessage = {
        id: generateId(),
        role: 'assistant',
        content: response.data.reply,
        timestamp: Date.now(),
      };
      messages.value.push(assistantMessage);
    } catch (error: any) {
      const errorMessage = error.response?.data?.detail || 'AI 对话服务出错';
      ElMessage.error(errorMessage);
      // Remove user message on error
      messages.value.pop();
    } finally {
      isLoading.value = false;
    }
  };

  const clearMessages = () => {
    messages.value = [];
  };

  /**
   * Parse JSON from AI response and return a list of field info for preview.
   */
  const getApplyFieldsInfo = (jsonContent: Record<string, any>): ApplyFieldInfo[] => {
    const fieldMap: Record<string, { path: string; label: string }> = {
      name:                       { path: 'data.name',                       label: 'Name / 名称' },
      description:                { path: 'data.description',                label: 'Description / 描述' },
      personality:                { path: 'data.personality',                label: 'Personality / 性格' },
      scenario:                   { path: 'data.scenario',                   label: 'Scenario / 场景' },
      first_mes:                  { path: 'data.first_mes',                  label: 'First Message / 第一条消息' },
      mes_example:                { path: 'data.mes_example',                label: 'Example Messages / 示例对话' },
      creator_notes:              { path: 'data.creator_notes',              label: 'Creator Notes / 创作者笔记' },
      system_prompt:              { path: 'data.system_prompt',              label: 'System Prompt / 系统提示' },
      post_history_instructions:  { path: 'data.post_history_instructions',  label: 'Post History / 历史指令' },
    };

    const fields: ApplyFieldInfo[] = [];

    for (const [key, meta] of Object.entries(fieldMap)) {
      const val = jsonContent[key];
      if (val !== undefined && val !== null && val !== '') {
        const strVal = typeof val === 'string' ? val : JSON.stringify(val);
        fields.push({
          key,
          path: meta.path,
          label: meta.label,
          value: val,
          preview: strVal.length > 60 ? strVal.slice(0, 60) + '...' : strVal,
        });
      }
    }

    // Handle tags
    if (jsonContent.tags && Array.isArray(jsonContent.tags)) {
      fields.push({
        key: 'tags',
        path: 'data.tags',
        label: 'Tags / 标签',
        value: jsonContent.tags,
        preview: jsonContent.tags.join(', '),
      });
    }

    // Handle alternate_greetings
    if (jsonContent.alternate_greetings && Array.isArray(jsonContent.alternate_greetings) && jsonContent.alternate_greetings.length > 0) {
      const greetings = jsonContent.alternate_greetings.filter((g: any) => typeof g === 'string' && g.trim());
      if (greetings.length > 0) {
        const previewText = greetings.length === 1
          ? greetings[0]
          : `${greetings.length} 条备用问候语`;
        fields.push({
          key: 'alternate_greetings',
          path: 'data.alternate_greetings',
          label: 'Alternate Greetings / 备用问候语',
          value: greetings,
          preview: previewText.length > 60 ? previewText.slice(0, 60) + '...' : previewText,
        });
      }
    }

    return fields;
  };

  /**
   * Apply selected fields to character card. Returns count of applied fields.
   */
  const applyToCard = (jsonContent: Record<string, any>, selectedKeys?: string[]) => {
    const translatorStore = useTranslatorStore();
    if (!translatorStore.characterCard) {
      translatorStore.createNewCard();
    }

    const allFields = getApplyFieldsInfo(jsonContent);
    const fieldsToApply = selectedKeys
      ? allFields.filter(f => selectedKeys.includes(f.key))
      : allFields;

    let appliedCount = 0;
    for (const field of fieldsToApply) {
      translatorStore.updateCardField(field.path, field.value);
      appliedCount++;
    }

    return appliedCount;
  };

  /**
   * Apply a single field to the card by key.
   */
  const applySingleField = (field: ApplyFieldInfo) => {
    const translatorStore = useTranslatorStore();
    if (!translatorStore.characterCard) {
      translatorStore.createNewCard();
    }
    translatorStore.updateCardField(field.path, field.value);
  };

  /**
   * Extract JSON blocks from AI response text.
   */
  const extractJsonFromResponse = (text: string): Record<string, any> | null => {
    // Try to find JSON block in markdown code fence
    const jsonMatch = text.match(/```(?:json)?\s*\n?([\s\S]*?)\n?```/);
    if (jsonMatch) {
      try {
        return JSON.parse(jsonMatch[1].trim());
      } catch (e) {
        // Not valid JSON
      }
    }

    // Try to parse the whole text as JSON
    try {
      return JSON.parse(text.trim());
    } catch (e) {
      // Not valid JSON
    }

    return null;
  };

  // --- Watchers ---
  watch(messages, (val) => {
    localStorage.setItem(CHAT_STORAGE_KEY, JSON.stringify(val));
  }, { deep: true });

  watch(sidebarVisible, (val) => {
    localStorage.setItem(CHAT_SIDEBAR_VISIBLE_KEY, String(val));
  });

  // --- Initial Load ---
  loadFromStorage();

  return {
    messages,
    isLoading,
    sidebarVisible,
    toggleSidebar,
    sendMessage,
    clearMessages,
    applyToCard,
    applySingleField,
    getApplyFieldsInfo,
    extractJsonFromResponse,
  };
});

File: src/routers/ai_chat.py
"""
AI 辅助角色卡生成对话路由
"""
import json
import logging

from fastapi import APIRouter, HTTPException

from ..models.schemas import AIChatRequest, AIChatResponse

router = APIRouter(prefix="/api/v1", tags=["ai-chat"])
logger = logging.getLogger(__name__)

# ------------------------------------------------------------------
# 系统提示词（从路由函数中分离为常量）
# ------------------------------------------------------------------

AI_CHAT_SYSTEM_PROMPT = """你是一个专业的 SillyTavern 角色卡创建助手。你的任务是帮助用户创建和完善 AI 角色卡。

角色卡包含以下字段：
- name: 角色名称
- description: 角色描述（外貌、背景、性格特征等详细设定）
- personality: 性格特征（简要概括）
- scenario: 场景设定（故事发生的背景）
- first_mes: 第一条消息（角色的开场白）
- alternate_greetings: 备用问候语（角色的备选开场白列表，是一个字符串数组）
- mes_example: 示例对话（展示角色的说话风格和互动方式）
- creator_notes: 创作者笔记
- system_prompt: 系统提示词（给 AI 的角色扮演指令）
- post_history_instructions: 历史指令（附加在对话历史后的指令）
- tags: 标签

你应该：
1. 根据用户的描述，帮助生成角色卡的各个字段内容
2. 提供创意建议和改进意见
3. 确保角色设定的一致性和丰富度
4. 当用户提供模糊的想法时，帮助他们细化和具象化
5. 可以一次性生成完整的角色卡，也可以逐字段讨论完善

当用户要求生成角色卡内容时，请使用以下 JSON 格式输出（可以只输出需要的字段）：
```json
{
  "name": "角色名",
  "description": "详细描述",
  "personality": "性格特征",
  "scenario": "场景设定",
  "first_mes": "第一条消息",
  "alternate_greetings": ["备用问候语1", "备用问候语2"],
  "mes_example": "示例对话",
  "creator_notes": "创作者笔记",
  "system_prompt": "系统提示词",
  "post_history_instructions": "历史指令",
  "tags": ["标签1", "标签2"]
}
```

在对话中要友好、专业，并根据用户的语言（中文或英文）来回复。"""


@router.post("/character/ai-chat", response_model=AIChatResponse)
async def ai_chat(data: AIChatRequest):
    """
    AI 辅助角色卡生成对话接口。
    接收对话历史和当前角色卡上下文，返回 AI 回复以帮助生成/完善角色卡。
    """
    if not data.settings.api_key:
        raise HTTPException(status_code=400, detail="请先在设置中提供您的 API Key。")

    try:
        from langchain_openai import ChatOpenAI
        from pydantic import SecretStr
        from langchain_core.messages import (
            SystemMessage as LCSystemMessage,
            HumanMessage,
            AIMessage,
        )

        llm = ChatOpenAI(
            model=data.settings.model_name,
            base_url=data.settings.base_url,
            api_key=SecretStr(data.settings.api_key),
            max_completion_tokens=8192,
        )

        # 构建系统提示词
        system_prompt = AI_CHAT_SYSTEM_PROMPT

        if data.character_card and isinstance(data.character_card, dict):
            card_data = data.character_card.get('data', data.character_card)
            card_info = json.dumps(card_data, ensure_ascii=False, indent=2)
            system_prompt += (
                f"\n\n当前角色卡数据：\n```json\n{card_info}\n```\n"
                "请基于上述现有数据为用户提供建议和帮助。"
            )

        # 构建 LangChain 消息列表
        lc_messages = [LCSystemMessage(content=system_prompt)]

        for msg in data.messages:
            if msg.role == 'user':
                lc_messages.append(HumanMessage(content=msg.content))
            elif msg.role == 'assistant':
                lc_messages.append(AIMessage(content=msg.content))

        response = llm.invoke(lc_messages)

        ai_content = response.content if isinstance(response.content, str) else str(response.content)

        return AIChatResponse(reply=ai_content)

    except Exception as e:
        logger.error(f"AI 对话过程中发生错误：{e}")
        error_message = str(e)
        if "auth" in error_message.lower() or "api key" in error_message.lower():
            raise HTTPException(status_code=401, detail="API Key 无效或已过期。")
        elif "rate" in error_message.lower():
            raise HTTPException(status_code=429, detail="请求过于频繁，请稍后再试。")
        raise HTTPException(
            status_code=500, detail=f"AI 对话过程中发生错误：{error_message}"
        )
